File: dotnet/Query/Repository.cs
#nullable enable
using Io.Cucumber.Messages.Types;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System;
using System.Linq;

namespace Io.Cucumber.Query
{
    public class Repository
    {
        // Features
        public enum RepositoryFeature
        {
            /// <summary>
            /// Include <see cref="Attachment"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_ATTACHMENTS,

            /// <summary>
            /// Include <see cref="Io.Cucumber.Messages.Types.GherkinDocument"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_GHERKIN_DOCUMENTS,

            /// <summary>
            /// Include <see cref="Hook"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_HOOKS,

            /// <summary>
            /// Include <see cref="StepDefinition"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_STEP_DEFINITIONS,

            /// <summary>
            /// Include <see cref="Suggestion"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_SUGGESTIONS,

            /// <summary>
            /// Include <see cref="UndefinedParameterType"/> messages.
            /// Disable to reduce memory usage.
            /// </summary>
            INCLUDE_UNDEFINED_PARAMETER_TYPES,
        }

        private readonly HashSet<RepositoryFeature> _features;

        // Public fields (matching Java order)
        public readonly Dictionary<string, TestCaseStarted> TestCaseStartedById = new();
        public readonly Dictionary<string, TestCaseFinished> TestCaseFinishedByTestCaseStartedId = new();
        public readonly Dictionary<string, List<TestStepFinished>> TestStepsFinishedByTestCaseStartedId = new();
        public readonly Dictionary<string, List<TestStepStarted>> TestStepsStartedByTestCaseStartedId = new();
        public readonly Dictionary<string, TestRunHookStarted> TestRunHookStartedById = new();
        public readonly Dictionary<string, TestRunHookFinished> TestRunHookFinishedByTestRunHookStartedId = new();
        public readonly Dictionary<string, Pickle> PickleById = new();
        public readonly Dictionary<string, TestCase> TestCaseById = new();
        public readonly Dictionary<string, Step> StepById = new();
        public readonly Dictionary<string, TestStep> TestStepById = new();
        public readonly Dictionary<string, PickleStep> PickleStepById = new();
        public readonly Dictionary<string, Hook> HookById = new();
        public readonly Dictionary<string, List<Attachment>> AttachmentsByTestCaseStartedId = new();
        public readonly Dictionary<string, List<Attachment>> AttachmentsByTestRunHookStartedId = new();
        public readonly Dictionary<object, Lineage> LineageById = new();
        public readonly Dictionary<string, StepDefinition> StepDefinitionById = new();
        public readonly Dictionary<string, List<Suggestion>> SuggestionsByPickleStepId = new();
        public readonly List<UndefinedParameterType> UndefinedParameterTypes = new();

        public Meta? Meta;
        public TestRunStarted? TestRunStarted;
        public TestRunFinished? TestRunFinished;

        public static Repository CreateWithAllFeatures()
        {
            return new Repository((RepositoryFeature[])Enum.GetValues(typeof(RepositoryFeature)));
        }

        public Repository(IEnumerable<RepositoryFeature>? features = null)
        {
            _features = features != null
                ? new HashSet<RepositoryFeature>(features)
                : new HashSet<RepositoryFeature>();
        }

        public void Update(Envelope envelope)
        {
            if (envelope.Meta != null) UpdateMeta(envelope.Meta);
            if (envelope.TestRunStarted != null) UpdateTestRunStarted(envelope.TestRunStarted);
            if (envelope.TestRunFinished != null) UpdateTestRunFinished(envelope.TestRunFinished);
            if (envelope.TestRunHookStarted != null) UpdateTestRunHookStarted(envelope.TestRunHookStarted);
            if (envelope.TestRunHookFinished != null) UpdateTestRunHookFinished(envelope.TestRunHookFinished);
            if (envelope.TestCaseStarted != null) UpdateTestCaseStarted(envelope.TestCaseStarted);
            if (envelope.TestCaseFinished != null) UpdateTestCaseFinished(envelope.TestCaseFinished);
            if (envelope.TestStepStarted != null) UpdateTestStepStarted(envelope.TestStepStarted);
            if (envelope.TestStepFinished != null) UpdateTestStepFinished(envelope.TestStepFinished);
            if (envelope.Pickle != null) UpdatePickle(envelope.Pickle);
            if (envelope.TestCase != null) UpdateTestCase(envelope.TestCase);
            if (_features.Contains(RepositoryFeature.INCLUDE_GHERKIN_DOCUMENTS) && envelope.GherkinDocument != null) UpdateGherkinDocument(envelope.GherkinDocument);
            if (_features.Contains(RepositoryFeature.INCLUDE_STEP_DEFINITIONS) && envelope.StepDefinition != null) UpdateStepDefinition(envelope.StepDefinition);
            if (_features.Contains(RepositoryFeature.INCLUDE_HOOKS) && envelope.Hook != null) UpdateHook(envelope.Hook);
            if (_features.Contains(RepositoryFeature.INCLUDE_ATTACHMENTS) && envelope.Attachment != null) UpdateAttachment(envelope.Attachment);
            if (_features.Contains(RepositoryFeature.INCLUDE_SUGGESTIONS) && envelope.Suggestion != null) UpdateSuggestions(envelope.Suggestion);
            if (_features.Contains(RepositoryFeature.INCLUDE_UNDEFINED_PARAMETER_TYPES) && envelope.UndefinedParameterType != null) UpdateUndefinedParameterType(envelope.UndefinedParameterType);
        }

        internal void UpdateAttachment(Attachment attachment)
        {
            if (attachment.TestCaseStartedId != null)
            {
                if (!AttachmentsByTestCaseStartedId.TryGetValue(attachment.TestCaseStartedId, out var list))
                {
                    list = new List<Attachment>();
                    AttachmentsByTestCaseStartedId[attachment.TestCaseStartedId] = list;
                }
                list.Add(attachment);
            }
            if (attachment.TestRunHookStartedId != null)
            {
                if (!AttachmentsByTestRunHookStartedId.TryGetValue(attachment.TestRunHookStartedId, out var list))
                {
                    list = new List<Attachment>();
                    AttachmentsByTestRunHookStartedId[attachment.TestRunHookStartedId] = list;
                }
                list.Add(attachment);
            }
        }

        internal void UpdateHook(Hook hook)
        {
            HookById[hook.Id] = hook;
        }

        internal void UpdateTestCaseStarted(TestCaseStarted testCaseStarted)
        {
            TestCaseStartedById[testCaseStarted.Id] = testCaseStarted;
        }

        internal void UpdateTestCase(TestCase testCase)
        {
            TestCaseById[testCase.Id] = testCase;
            foreach (var testStep in testCase.TestSteps)
            {
                TestStepById[testStep.Id] = testStep;
            }
        }

        internal void UpdatePickle(Pickle pickle)
        {
            PickleById[pickle.Id] = pickle;
            foreach (var step in pickle.Steps)
            {
                PickleStepById[step.Id] = step;
            }
        }

        internal void UpdateGherkinDocument(GherkinDocument document)
        {
            foreach (var lineage in Lineages.Of(document))
            {
                LineageById.Add(lineage.Key, lineage.Value);
            }
            if (document.Feature != null)
            {
                UpdateFeature(document.Feature);
            }
        }

        internal void UpdateFeature(Feature feature)
        {
            foreach (var child in feature.Children)
            {
                if (child.Background != null)
                {
                    UpdateSteps(child.Background.Steps);
                }
                if (child.Scenario != null)
                {
                    UpdateScenario(child.Scenario);
                }
                if (child.Rule != null)
                {
                    foreach (var ruleChild in child.Rule.Children)
                    {
                        if (ruleChild.Background != null)
                        {
                            UpdateSteps(ruleChild.Background.Steps);
                        }
                        if (ruleChild.Scenario != null)
                        {
                            UpdateScenario(ruleChild.Scenario);
                        }
                    }
                }
            }
        }

        internal void UpdateTestStepStarted(TestStepStarted testStepStarted)
        {
            if (!TestStepsStartedByTestCaseStartedId.TryGetValue(testStepStarted.TestCaseStartedId, out var list))
            {
                list = new List<TestStepStarted>();
                TestStepsStartedByTestCaseStartedId[testStepStarted.TestCaseStartedId] = list;
            }
            list.Add(testStepStarted);
        }

        internal void UpdateTestStepFinished(TestStepFinished testStepFinished)
        {
            if (!TestStepsFinishedByTestCaseStartedId.TryGetValue(testStepFinished.TestCaseStartedId, out var list))
            {
                list = new List<TestStepFinished>();
                TestStepsFinishedByTestCaseStartedId[testStepFinished.TestCaseStartedId] = list;
            }
            list.Add(testStepFinished);
        }

        internal void UpdateTestCaseFinished(TestCaseFinished testCaseFinished)
        {
            TestCaseFinishedByTestCaseStartedId[testCaseFinished.TestCaseStartedId] = testCaseFinished;
        }

        internal void UpdateTestRunFinished(TestRunFinished testRunFinished)
        {
            TestRunFinished = testRunFinished;
        }

        internal void UpdateTestRunStarted(TestRunStarted testRunStarted)
        {
            TestRunStarted = testRunStarted;
        }

        internal void UpdateTestRunHookStarted(TestRunHookStarted testRunHookStarted)
        {
            TestRunHookStartedById[testRunHookStarted.Id] = testRunHookStarted;
        }

        internal void UpdateTestRunHookFinished(TestRunHookFinished testRunHookFinished)
        {
            TestRunHookFinishedByTestRunHookStartedId[testRunHookFinished.TestRunHookStartedId] = testRunHookFinished;
        }

        internal void UpdateScenario(Scenario scenario)
        {
            UpdateSteps(scenario.Steps);
        }

        internal void UpdateStepDefinition(StepDefinition stepDefinition)
        {
            StepDefinitionById[stepDefinition.Id] = stepDefinition;
        }

        internal void UpdateSteps(IList<Step> steps)
        {
            foreach (var step in steps)
            {
                StepById[step.Id] = step;
            }
        }

        internal void UpdateSuggestions(Suggestion suggestion)
        {
            if (!SuggestionsByPickleStepId.TryGetValue(suggestion.PickleStepId, out var list))
            {
                list = new List<Suggestion>();
                SuggestionsByPickleStepId[suggestion.PickleStepId] = list;
            }
            list.Add(suggestion);
        }

        internal void UpdateMeta(Meta meta)
        {
            Meta = meta;
        }

        internal void UpdateUndefinedParameterType(UndefinedParameterType undefinedParameterType)
        {
            UndefinedParameterTypes.Add(undefinedParameterType);
        }
    }
}

File: ruby/lib/cucumber/repository.rb
# frozen_string_literal: true

module Cucumber
  # In memory repository i.e. a thread based link to cucumber-query
  class Repository
    attr_accessor :meta, :test_run_started, :test_run_finished
    attr_reader :attachments_by_test_case_started_id, :attachments_by_test_run_hook_started_id,
                :hook_by_id,
                :pickle_by_id, :pickle_step_by_id,
                :suggestions_by_pickle_step_id,
                :step_by_id, :step_definition_by_id,
                :test_case_by_id, :test_case_started_by_id, :test_case_finished_by_test_case_started_id,
                :test_run_hook_started_by_id, :test_run_hook_finished_by_test_run_hook_started_id,
                :test_step_by_id, :test_steps_started_by_test_case_started_id, :test_steps_finished_by_test_case_started_id,
                :undefined_parameter_types

    def initialize
      @attachments_by_test_case_started_id = _hash_with_array_default
      @attachments_by_test_run_hook_started_id = _hash_with_array_default
      @hook_by_id = {}
      @pickle_by_id = {}
      @pickle_step_by_id = {}
      @step_by_id = {}
      @step_definition_by_id = {}
      @suggestions_by_pickle_step_id = _hash_with_array_default
      @test_case_by_id = {}
      @test_case_started_by_id = {}
      @test_case_finished_by_test_case_started_id = {}
      @test_run_hook_started_by_id = {}
      @test_run_hook_finished_by_test_run_hook_started_id = {}
      @test_step_by_id = {}
      @test_steps_started_by_test_case_started_id = _hash_with_array_default
      @test_steps_finished_by_test_case_started_id = _hash_with_array_default
      @undefined_parameter_types = []
    end

    def update(envelope)
      message_category = envelope.type[:contained_message]
      message_value = envelope.public_send(message_category)
      send("update_#{message_category}", message_value)
    end

    private

    def method_missing(method_name, *args, &)
      if method_name.to_s.start_with?('update_')
        Kernel.warn("Attempting to update the repository with #{method_name}. Please raise this as a missing message handler")
        Kernel.warn('Create an issue here: https://github.com/cucumber/query/issues')
        nil
      else
        super
      end
    end

    def respond_to_missing?(method_name, include_private = false)
      method_name.to_s.start_with?('update_') || super
    end

    def update_attachment(attachment)
      attachments_by_test_case_started_id[attachment.test_case_started_id] << attachment if attachment.test_case_started_id
      attachments_by_test_run_hook_started_id[attachment.test_run_hook_started_id] << attachment if attachment.test_run_hook_started_id
    end

    def update_feature(feature)
      feature.children.each do |feature_child|
        update_steps(feature_child.background.steps) if feature_child.background
        update_scenario(feature_child.scenario) if feature_child.scenario
        feature_child.rule&.children&.each { |rule_child| _update_feature_rule(rule_child) }
      end
    end

    def update_gherkin_document(gherkin_document)
      update_feature(gherkin_document.feature) if gherkin_document.feature
    end

    def update_hook(hook)
      hook_by_id[hook.id] = hook
    end

    def update_meta(meta)
      self.meta = meta
    end

    def update_pickle(pickle)
      pickle_by_id[pickle.id] = pickle
      pickle.steps.each { |pickle_step| pickle_step_by_id[pickle_step.id] = pickle_step }
    end

    def update_scenario(scenario)
      update_steps(scenario.steps)
    end

    def update_source(_source)
      # This deliberately doesn't perform any handling. `Source` as a message is not stored or required
      #   - See `GherkinDocument` for a more "parsed" form of an AST representation
      :no_op
    end

    def update_steps(steps)
      steps.each { |step| step_by_id[step.id] = step }
    end

    def update_step_definition(step_definition)
      step_definition_by_id[step_definition.id] = step_definition
    end

    def update_suggestion(suggestion)
      suggestions_by_pickle_step_id[suggestion.pickle_step_id] << suggestion
    end

    def update_test_case(test_case)
      test_case_by_id[test_case.id] = test_case
      test_case.test_steps.each { |test_step| test_step_by_id[test_step.id] = test_step }
    end

    def update_test_case_started(test_case_started)
      test_case_started_by_id[test_case_started.id] = test_case_started
    end

    def update_test_case_finished(test_case_finished)
      test_case_finished_by_test_case_started_id[test_case_finished.test_case_started_id] = test_case_finished
    end

    def update_test_run_started(test_run_started)
      self.test_run_started = test_run_started
    end

    def update_test_run_finished(test_run_finished)
      self.test_run_finished = test_run_finished
    end

    def update_test_run_hook_started(test_run_hook_started)
      test_run_hook_started_by_id[test_run_hook_started.id] = test_run_hook_started
    end

    def update_test_run_hook_finished(test_run_hook_finished)
      test_run_hook_finished_by_test_run_hook_started_id[test_run_hook_finished.test_run_hook_started_id] = test_run_hook_finished
    end

    def update_test_step_started(test_step_started)
      test_steps_started_by_test_case_started_id[test_step_started.test_case_started_id] << test_step_started
    end

    def update_test_step_finished(test_step_finished)
      test_steps_finished_by_test_case_started_id[test_step_finished.test_case_started_id] << test_step_finished
    end

    def update_undefined_parameter_type(undefined_parameter_type)
      undefined_parameter_types << undefined_parameter_type
    end

    def _update_feature_rule(rule_child)
      update_steps(rule_child.background.steps) if rule_child.background
      update_scenario(rule_child.scenario) if rule_child.scenario
    end

    def _hash_with_array_default
      Hash.new { |hash, key| hash[key] = [] }
    end
  end
end
